### ISSUE TITLE
Modifying length of file can break VCR integration

### ISSUE THREAD

**Issue by lkysow** [php-vcr/php-vcr#76]
**Tl;dr: Changing the length of your file by adding/removing code can break PHP VCR.**
### How The Code Works

PHP VCR modifies files when they're being read from the filesystem through the use of PHP's `php_user_filter` (http://php.net/manual/en/class.php-user-filter.php).

When a file is read from the filesystem, it goes through this function in PHP VCR

``` php
<?php
public function filter($in, $out, &$consumed, $closing)
{
        while ($bucket = stream_bucket_make_writeable($in)) {
            $bucket->data = $this->transformCode($bucket->data);
            $consumed += $bucket->datalen;
            stream_bucket_append($out, $bucket);
        }

        return PSFS_PASS_ON;
    }
}
```

which calls  `transformCode($code)` in each code transformer

``` php
<?php

class CurlCodeTransform extends AbstractCodeTransform
{

    private static $patterns = array(
        '/(?<!::|->)\\\?curl_init\s*\(/i' => '\VCR\LibraryHooks\CurlHook::curl_init('
        ...
    );

    protected function transformCode($code)
    {
        return preg_replace(array_keys(self::$patterns), array_values(self::$patterns), $code);
    }
}
```

The `transformCode($code)` function does a `preg_replace` and replaces the original code with code that calls PHP VCR.
### The Problem

You may have noticed that the `filter` function operates in a `while` loop calling `stream_bucket_make_writeable`. The bug is that this function can split up the stream, i.e. the file that is being read. 

If it splits it between a line of code that we're trying to search and replace, then the `preg_replace` will fail and we won't modify this file.
### Example

The class we're rewriting:

``` php
<?php
class TransformMe {
    public function test() {
        $curl = curl_init();
    }
}
```

Ideally, this will be re-written to 

``` php
<?php
class TransformMe {
    public function test() {
        $curl = \VCR\LibraryHooks\CurlHook::curl_init();
    }
}
```

However if the file is split like this:

``` php
<?php
class TransformMe {
    public function test() {
        $curl = curl_
```

``` php
init();
    }
}
```

Then the transformation will never happen.

**This was fun to debug :)**
### Solutions?

I'd say the best thing to do would be to read the entire file at once, and _then_ do the search and replace, rather than relying on PHP's streams and bucketing.

**Comment by higidi** [comment]
Closing — fixed on master.

`AbstractCodeTransform::filter()` now accumulates all stream buckets into `$this->data` and only calls `transformCode()` on `$closing`, so the entire file content is available for the regex replace before any rewriting happens. The split-bucket problem described here no longer occurs.

Fix was introduced via PR #77.

**Comment by lkysow** [comment]
Thanks for updating a 12 year old bug!
